### PR TITLE
Add async support

### PR DESCRIPTION
**Pull request by Tim-Zhang**
Add supports for async/.await.

By using async/.await you can reduce the overhead and resource consumption caused by threads.

Signed-off-by: Tim Zhang <tim@hyper.sh>

### CODE REVIEW

File: src/asynchronous/client.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            // notify the response-receiver to quit at this time.
+            notify.notify();
+        });
+
+        let req_map2 = req_map.clone();
+
+        // Response receiver
+        tokio::spawn(async move {
+            loop {
+                let req_map2 = req_map2.clone();
```
Comment (lifupan) [inline src/asynchronous/client.rs]: can the req_map moved here instead of clone?

File: src/asynchronous/client.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                                    {
+                                        let mut map = req_map2.lock().unwrap();
+                                        let resp_tx = match map.get(&header.stream_id) {
+                                            Some(tx) => tx,
+                                            None => {
+                                                debug!(
+                                                    "Receiver got unknown packet {:?} {:?}",
+                                                    header, body
+                                                );
+                                                return;
```
Comment (lifupan) [inline src/asynchronous/client.rs:105]: is it better to skip this unexpected pkg instead of return?

File: src/asynchronous/client.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                let current_stream_id = stream_id;
+                stream_id += 2;
+
+                {
+                    let mut map = req_map2.lock().unwrap();
+                    map.insert(current_stream_id, resp_tx.clone());
+                }
+
+                let buf = to_req_buf(current_stream_id, body);
+                if let Err(e) = writer.write_all(&buf).await {
```
Comment (lifupan) [inline src/asynchronous/client.rs:65]: if write failed， which means the connection was broken, should it break the loop and notify receiver to terminate.

Comment (Tim-Zhang) [reply src/asynchronous/client.rs:65]: done

**Pull request merged**